utf8/reader_test.go

```diff
@@ -0,0 +1,288 @@
+package utf8_test
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"testing"
+
+	stdiotest "testing/iotest"
+
+	"github.com/madlambda/spells/assert"
+	"github.com/madlambda/spells/io/runes"
+	"github.com/madlambda/spells/iotest"
+	"github.com/madlambda/spells/utf8"
+)
+
+const expectedString = "Εν οίδα ότι ουδέν οίδα"
+
+func TestUTF8Reader(t *testing.T) {
+	type testcase struct {
+		input  string
+		repeat int
+		err    error
+		errOut []rune
+	}
+
+	testcases := []testcase{
+		{
+			input:  "",
+			repeat: 1,
+		},
+		{
+			input:  string([]byte{0x80}),
+			repeat: 1,
+			err:    utf8.ErrDecode,
+		},
+		{
+			input:  "test",
+			repeat: 1,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 1,
+		},
+		{
+			input:  "Εν οίδα ότι TESTE ουδέν οίδα",
+			repeat: 1,
+		},
+		{
+			input:  "test\n",
+			repeat: 1024,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 1024,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 1025,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 2048,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 2049,
+		},
+		{
+			input:  string([]byte{206, 149, 206, 206, 206}),
+			repeat: 1,
+			err:    utf8.ErrDecode,
+			errOut: []rune{'Ε'},
+		},
+	}
+
+	// test NewReaderReader
```

Comment:
Hmmm not so sure about test NewReaderReader being together with NewReader in a way that if NewReaderReader test fails it will abort and not test NewReader. We could split into actual separate test functions or we could use t.Run here, avoiding the comment + adding isolation to failre on each major test, like:

```go
t.Run("ReaderReader", func(t *testing.t) {
     // the whole code doing the test
})
```

I'm not sure yet, but the testing itself also seems quite duplicated between the ReaderReader and the Reader (renaming these would be great, Im quite confused 😆 ). So maybe we could have a function that given a test case + the already built object that implements the interface, it tests it. At first it seems possible, but I could be missing something.

If the duplication is necessary (not duplicated after all), I would at least isolate the tests from each other (with t.Run at least). You can also leverage some parallel testing with t.Run...but overall the main issue is isolation on failure (one type failing a test should not stop me from testing a completely different type).